tx-parser/src/lib.rs

```diff
@@ -106,18 +121,23 @@ def_matches_namespaced_keyword!(Tx, literal_db_retract, "db", "retract");
 
 def_parser!(Tx, add_or_retract, Entity, {
     vector().of_exactly(
+        // TODO: This commits as soon as it sees :db/{add,retract}, but could use an improved error message.
         (Tx::literal_db_add().map(|_| OpType::Add).or(Tx::literal_db_retract().map(|_| OpType::Retract)),
-             Tx::entid_or_lookup_ref_or_temp_id(),
-             Tx::entid(),
-             Tx::atom_or_lookup_ref_or_vector())
-        .map(|(op, e, a, v)| {
-            Entity::AddOrRetract {
-                op: op,
-                e: e,
-                a: a,
-                v: v,
-            }
-        }))
+          try((Tx::entid_or_lookup_ref_or_temp_id(),
+               Tx::forward_entid(),
+               Tx::atom_or_lookup_ref_or_vector()))
+          .or(try((Tx::atom_or_lookup_ref_or_vector(),
```

Comment:
I think you're insticts are correct.  I thought for a short time about this, and I don't know how to achieve better with `combine`; and the first line of the `combine` README says it's LL(1) unless you opt-in to more lookahead with `try`, so this is the blessed way.

The best way might be to parse `:db/{add,retract} edn::Value :attribute/{forward,backward} edn::Value` and then to parse the two `edn::Value` pieces into the correct generalities.  Otherwise you don't get the restriction on the entid location that you want in the parser, which is how we reject mal-formed input early.

My expectation is that `try` is really pretty fast in this situation, especially 'cuz backward entities are rare.